File: src/backend/kmp.go
package main

import "fmt"

func kmpMatch(text string, pattern string) int {
    n := len(text)
    m := len(pattern)

    lps := computeLPSArray(pattern)

    i := 0
    j := 0
    for i < n {
        if pattern[j] == text[i] {
            if j == m-1 {
                return i - m + 1
            }
            i++
            j++
        } else if j > 0 {
            j = lps[j-1]
        } else {
            i++
        }
    }
    return -1
}

func computeLPSArray(pattern string) []int {
    m := len(pattern)
    lps := make([]int, m)

    j := 0
    i := 1
    for i < m {
        if pattern[i] == pattern[j] {
            j++
            lps[i] = j
            i++
        } else if j > 0 {
            j = lps[j-1]
        } else {
            lps[i] = 0
            i++
        }
    }
    return lps
}

func main() {
    text := "ABABDABACDABABCABAB"
    pattern := "ABABDABACDABABCABAB"
    fmt.Println(kmpMatch(text, pattern))
}

File: src/backend/kmp.js
function kmpMatch(text, pattern){
    var n = text.length;
    var m = pattern.length;

    var lps = computeLPSArray(pattern);

    var i = 0;
    var j = 0;
    while(i < n){
        if(pattern[j] == text[i]){
            if (j == m - 1){
                return i - m + 1;
            }
            i++;
            j++;
        }
        else if (j > 0){
            j = lps[j-1];

        }
        else{
            i++;
        }
    }
    return -1;

}

function computeLPSArray(pattern){
    var lps = new Array(pattern.length);
    var m = pattern.length;
    var j = 0;
    var i = 1;

    while(i < m){
        if(pattern[i] == pattern[j]){
            j++;
            lps[i] = j;
            i++;
            j++;

        }
        else if(j > 0){
            j = lps[j-1];
        }
        else{
            lps[i] = 0;
            i++;
        }
    }
    return lps;

}

function main(){
    var text = "ABABDABACDABABCABAB";
    var pattern = "ABABDABCDABABCABAB";
    console.log(kmpMatch(text, pattern))
}
main();